### ISSUE TITLE
Towards a better (systematic and automatized) index of error/warning messages in the reference manual?

### ISSUE THREAD

**Issue by herbelin** [coq/coq#7373]
I'm copying here a proposal made during a discussion started at #7315.

@zimmi48: The thing is we currently provide an error message index: https://coq.inria.fr/refman/coq-exnindex.html. Users can use it to better understand error messages they get. Currently the index is incomplete. What I propose is just to make it more complete. But anyways, I won't bother you with this any longer.

@herbelin: That's a good point.

A quick grep gives 72 `CWarnings.create`, about 900 `user_err`, 190 extra errors centralized in `himsg.ml`, and about 200 errors or warnings in the index. So a ratio of 1/6, though probably among the most standard and thus useful ones.

So, we can maybe try to make progresses. I don't know how we could organize but the experience showed that we can achieve big tasks collectively. So, let's assume that we add an argument to each `user_err` to tell to which command it pertains, that we add systematically in the source of the manual a link to a section listing the errors attached to each command, that we have a tool to automatically translate such `user_err` into a list of errors for the given command in the manual, and, finally, that we share the work for indexing all such `user_err`. Do you think it is feasible?

Additionally, that would help us to have more uniform error messages.

@Zimmi48: About the project you propose: it is an interesting one and I can't assess its feasibility. I guess it would at least need to first wait for the completion of the error message handling redesign. I suggest we (you?) open a new issue about this (as food for thought for the moment).

**Comment by ejgallego** [comment]
ccing @cpitclaudel as he has requested a more systematic error index a few times [and I'm sure he wrote about it around here]

**Comment by cpitclaudel** [comment]
Yup, a more systematic error index would be a bliss.
Right now company-coq takes all the words in the error message, sorts them, and uses a simple distance metric to match it against error messages in the refman.

**Comment by ejgallego** [comment]
Hi folks, see #7560 as a first steps towards this. In fact I am optimistic we will be able to access list of warnings / errors from shells / SerAPI soon.

**Comment by ejgallego** [comment]
Then the open question for the documentation folks is what kind of metadata you want to add to error / warning registration and what kind of format do you need the registration modules to output.

A slightly tricky problem with this approach, is that in order to obtain errors for plugins the plugin itself must be loaded. I don't think that's a big problem but certainly adds some annoyances to the doc build process.

**Comment by Zimmi48** [comment]
> A slightly tricky problem with this approach, is that in order to obtain errors for plugins the plugin itself must be loaded. I don't think that's a big problem but certainly adds some annoyances to the doc build process.

Well, because of coqtop snippets the documentation currently depends on a full Coq build anyways, so plugins are already built and loaded I think...

**Comment by ejgallego** [comment]
Yeah but I guess at some point you want to do a global "dump", certainly plugin looks like it should be a category. I guess we will see what we need when we reach there.
